conformance/README.md

```diff
@@ -172,66 +172,106 @@ ensuring that all input forms handle scenarios the same way.
 
 ### Combining Formats
 
-TODO: What does combining text + binary mean?
+The previous section implies that in general, a test case can be thought of as a
+tree where the interior nodes are fragments, and the leaves are expectations.
+Each expectation is tied to a single document formed by combining the fragments
+along the path from the root to the expectation.
 
-In theory, we can mix and match text, binary, and AST in a single
+In theory, we could mix and match text, binary, and AST fragments in a single
 document, since ultimately they all express the same thing in different ways.
-This means the test-runner cannot simply concatenate fragments, and it must
-either switch parsing modes mid-stream (theoretically sound since fragments must
-contain full top-level values), or transcribe the fragment into one format.
-But there are benefits to this approach:
+However, that forces fairly complicated transcoding into the test framework.
+To keep things reasonable, we constrain test trees such that `text` and `binary`
+fragments cannot coexist on the path to one expectation.
+AST fragments, however, can mix with text or binary fragments.
+This is the case in most situations, since the `ion_1_*` clauses inherently
+abstract over the bytes on an IVM.
+
+Before checking an expectation, the conformance framework constructs a document
+by combining the preceding sequence of fragments:
+  * If any fragment is `text`, any AST fragments are effectively converted to
+text before processing.  Adjacent text fragments MUST be joined with whitespace.
+  * If any fragment is `binary`, any AST fragments are effectively encoded to
+binary before processing.
```

Comment:
As the fragments are being traversed, the Ion encoding is always knowable.  At `document` it's 1.0 by default.  After each fragment it's still known, or at least known to the writer and an implementation could (if not should) surface it via the writing APIs.  I would like to assume that the framework can leverage that.

But I'm also not sure why the test framework needs to care; the "which version" is known to the writer it's transcribing into.

Probably it goes without saying but I think that both `text` and `bytes` should be handled, as much as possible, as direct byte-copy.  `byte`s in particular should not be modified during document aggregation; that would defeat the purpose.

Reply:
Hm, I wonder if the issue is that the writing APIs assume that the Ion-version is determined before creation, and cannot be changed mid-stream.  I would call that an insufficient API, since the Ion format allows version-switching in the midst of a document/stream, so the writers should allow that as well.

Reply:
> At document it's 1.0 by default

I think that this is enough to answer my original question.

> both text and bytes should be handled, as much as possible, as direct byte-copy

Naturally.

>  I would call that an insufficient API, since the Ion format allows version-switching in the midst of a document/stream, so the writers should allow that as well.

If you want to switch versions mid-stream while writing, `ion-java` currently requires you to create a new writer and append the second stream to the first one. We think this is sufficient because most real-world use cases probably don't need to _write_ multiple versions to one stream. Our reader _can_ switch versions mid stream.

I can figure out how to make it work with IVMs that occur in the middle of a fragment.